spatial/r3/transform.go

```diff
@@ -0,0 +1,267 @@
+func (t Transform) Transform(v Vec) Vec {
+	// https://github.com/mrdoob/three.js/blob/dev/src/math/Vector3.js#L262
+	w := 1 / (t.x30*v.X + t.x31*v.Y + t.x32*v.Z + t.d33 + 1)
+	return Vec{
+		X: ((t.d00+1)*v.X + t.x01*v.Y + t.x02*v.Z + t.x03) * w,
+		Y: (t.x10*v.X + (t.d11+1)*v.Y + t.x12*v.Z + t.x13) * w,
+		Z: (t.x20*v.X + t.x21*v.Y + (t.d22+1)*v.Z + t.x23) * w,
+	}
+}
+
+// zeroTransform is the Transform that returns zeroTransform when multiplied by any Transform.
+var zeroTransform = Transform{d00: -1, d11: -1, d22: -1, d33: -1}
+
+// NewTransform returns a new Transform type and populates its elements
+// with values passed in row-major form. If val is nil then NewTransform
+// returns a Transform filled with zeros.
+func NewTransform(a []float64) Transform {
+	if a == nil {
+		return zeroTransform
+	}
+	if len(a) != 16 {
+		panic("Transform is initialized with 16 values")
+	}
+	return Transform{
+		d00: a[0] - 1, x01: a[1], x02: a[2], x03: a[3],
+		x10: a[4], d11: a[5] - 1, x12: a[6], x13: a[7],
+		x20: a[8], x21: a[9], d22: a[10] - 1, x23: a[11],
+		x30: a[12], x31: a[13], x32: a[14], d33: a[15] - 1,
+	}
+}
+
+// ComposeTransform creates a new transform for a given translation to
+// positon, scaling vector scale and quaternion rotation.
+// The identity Transform is constructed with
+//  ComposeTransform(Vec{}, Vec{1,1,1}, Rotation{})
+func ComposeTransform(position, scale Vec, q Rotation) Transform {
+	x2 := q.Imag + q.Imag
+	y2 := q.Jmag + q.Jmag
+	z2 := q.Kmag + q.Kmag
+	xx := q.Imag * x2
+	yy := q.Jmag * y2
+	zz := q.Kmag * z2
+	xy := q.Imag * y2
+	xz := q.Imag * z2
+	yz := q.Jmag * z2
+	wx := q.Real * x2
+	wy := q.Real * y2
+	wz := q.Real * z2
+
+	var t Transform
+	t.d00 = (1-(yy+zz))*scale.X - 1
+	t.x10 = (xy + wz) * scale.X
+	t.x20 = (xz - wy) * scale.X
+
+	t.x01 = (xy - wz) * scale.Y
+	t.d11 = (1-(xx+zz))*scale.Y - 1
+	t.x21 = (yz + wx) * scale.Y
+
+	t.x02 = (xz + wy) * scale.Z
+	t.x12 = (yz - wx) * scale.Z
+	t.d22 = (1-(xx+yy))*scale.Z - 1
+
+	t.x03 = position.X
+	t.x13 = position.Y
+	t.x23 = position.Z
+	return t
+}
+
+// Translate adds Vec to the positional Transform.
+func (t Transform) Translate(v Vec) Transform {
+	t.x03 += v.X
+	t.x13 += v.Y
+	t.x23 += v.Z
+	return t
+}
+
+// Scale returns the transform with scaling added around
+// the argumnt origin.
+func (t Transform) Scale(origin, factor Vec) Transform {
+	if origin == (Vec{}) {
+		return t.scale(factor)
+	}
+	t = t.Translate(Scale(-1, origin))
+	t = t.scale(factor)
+	return t.Translate(origin)
+}
+
+func (t Transform) scale(factor Vec) Transform {
+	t.d00 = (t.d00+1)*factor.X - 1
+	t.x10 *= factor.X
+	t.x20 *= factor.X
+	t.x30 *= factor.X
+
+	t.x01 *= factor.Y
+	t.d11 = (t.d11+1)*factor.Y - 1
+	t.x21 *= factor.Y
+	t.x31 *= factor.Y
+
+	t.x02 *= factor.Z
+	t.x12 *= factor.Z
+	t.d22 = (t.d22+1)*factor.Z - 1
+	t.x32 *= factor.Z
+	return t
+}
+
+// Mul multiplies the Transforms a and b and returns the result.
+// This is the equivalent of combining two transforms in one.
```

Comment:
I think this could be clearer. The text as it stands does not indicate that the operation is non-commutative.